cmd/vspd/main.go

```diff
@@ -28,122 +30,143 @@ const maxVoteChangeRecords = 10
+		os.Exit(1)
+	}
+
+	vspd, err := newVspd(cfg)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "newVspd error: %v\n", err)
+		os.Exit(1)
+	}
+
 	// Run until an exit code is returned.
-	os.Exit(run())
+	os.Exit(vspd.run())
 }
 
-// run is the main startup and teardown logic performed by the main package. It
-// is responsible for parsing the config, creating dcrd and dcrwallet RPC clients,
-// opening the database, starting the webserver, and stopping all started
-// services when a shutdown is requested.
-func run() int {
+type vspd struct {
+	cfg         *config
+	log         slog.Logger
+	shutdownCtx context.Context
+	db          *database.VspDatabase
+	dcrd        rpc.DcrdConnect
+	wallets     rpc.WalletConnect
+}
 
-	// Load config file and parse CLI args.
-	cfg, err := loadConfig()
+// newVspd creates the essential resources required by vspd - a database, logger
+// and RPC clients - then returns an instance of vspd which is ready to be run.
+func newVspd(cfg *config) (*vspd, error) {
+
+	// Open database.
+	db, err := database.Open(cfg.dbPath, cfg.logger(" DB"), maxVoteChangeRecords)
 	if err != nil {
-		fmt.Fprintf(os.Stderr, "Config error: %v\n", err)
-		return 1
+		return nil, fmt.Errorf("failed to open database: %w", err)
 	}
 
 	log := cfg.logger("VSP")
-	dbLog := cfg.logger(" DB")
-	apiLog := cfg.logger("API")
 	rpcLog := cfg.logger("RPC")
 
-	// Show version at startup.
-	log.Criticalf("Version %s (Go version %s %s/%s)", version.String(), runtime.Version(),
-		runtime.GOOS, runtime.GOARCH)
+	// Create RPC client for local dcrd instance (used for broadcasting and
+	// checking the status of fee transactions).
+	dcrd := rpc.SetupDcrd(cfg.DcrdUser, cfg.DcrdPass, cfg.DcrdHost, cfg.dcrdCert, cfg.netParams.Params, rpcLog)
 
-	if cfg.netParams == &mainNetParams &&
-		version.PreRelease != "" {
-		log.Warnf("")
-		log.Warnf("\tWARNING: This is a pre-release version of vspd which should not be used on mainnet.")
-		log.Warnf("")
-	}
+	// Create RPC client for remote dcrwallet instances (used for voting).
+	wallets := rpc.SetupWallet(cfg.walletUsers, cfg.walletPasswords, cfg.walletHosts, cfg.walletCerts, cfg.netParams.Params, rpcLog)
 
-	if cfg.VspClosed {
-		log.Warnf("")
-		log.Warnf("\tWARNING: Config --vspclosed is set. This will prevent vspd from accepting new tickets.")
-		log.Warnf("")
+	v := &vspd{
+		cfg:         cfg,
+		log:         log,
+		shutdownCtx: shutdownListener(log),
+		db:          db,
+		dcrd:        dcrd,
+		wallets:     wallets,
 	}
 
-	defer log.Criticalf("Shutdown complete")
+	return v, nil
+}
 
-	// Open database.
-	db, err := database.Open(cfg.dbPath, dbLog, maxVoteChangeRecords)
-	if err != nil {
-		log.Errorf("Database error: %v", err)
-		return 1
+// run starts all of vspds background services including the web server, and
+// stops all started services when a shutdown is requested.
+func (v *vspd) run() int {
+
+	v.log.Criticalf("Version %s (Go version %s %s/%s)", version.String(), runtime.Version(),
```

Comment:
Critical is supposed to be for fatal failures.  Most libs even do full on aborts `os.Exit` style on critical.  For example, [see how wallet only uses it with the Fatal family of funcs when implementing the stdlib interface](https://github.com/decred/dcrwallet/blob/a87fa843495ec57c1d3b478c2ceb3876c3749af5/internal/rpc/rpcserver/log.go#L56).

I know the code was already this way, but it's not really correct.